### PR TITLE
Enable blueprint to define custom subgenerator

### PR DESCRIPTION
**Pull request by sahbi-ktifa**
Blueprints can easily modify main generator behavior, however, quickly grows the need to define custom subgenerator for specific tasks.

To ensure that possibility, this PR enables to define a subgenerator in a blueprint using a simple file formatted as such:
*generator-jhipster-testblueprint/commands.js*
```
module.exports = {
    foo: {
        desc: 'bar', // A little description that will be displayed using --help
        blueprint: 'generator-jhipster-testblueprint' // This is required to determine the blueprint issuing the custom command
    }
};
```

When using `jhipster foo --blueprint testblueprint`, the foo subgenerator from the blueprint will be called directly.

I'd like some reviews on this one, as on the feature itself as on the code. This will clearly enable users to define specific stuff on their own enjoying all JHipster cli mechanism.

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by pascalgrimaud** [comment]
Just to know, is it doable to add a test in this folder ?
https://github.com/jhipster/generator-jhipster/tree/master/test/blueprint

**Comment by sahbi-ktifa** [comment]
I added a test to this !
I can only test that it will try to delegate its job to the blueprint directly :)

**Comment by deepu105** [comment]
Let me take a look at this

File: cli/utils.js [hunk 1]
```diff
@@ -188,7 +188,12 @@ const createYeomanEnv = () => {
     Object.keys(SUB_GENERATORS)
         .filter(command => !SUB_GENERATORS[command].cliOnly)
         .forEach(generator => {
-            env.register(require.resolve(`../generators/${generator}`), `${CLI_NAME}:${generator}`);
+            if (SUB_GENERATORS[generator].blueprint) {
+                /* eslint-disable prettier/prettier */
+                env.register(require.resolve(`../node_modules/${SUB_GENERATORS[generator].blueprint}/generators/${generator}`), `${CLI_NAME}:${generator}`);
```
Comment (deepu105) [inline cli/utils.js]: this expects the package to be installed on the node_modules of the generator? Shouldn't you rely on npms resolution mechanism here

File: cli/commands.js [hunk 2]
```diff
@@ -103,3 +111,4 @@ Example:
         desc: 'Upgrade the JHipster version, and upgrade the generated application'
     }
 };
+module.exports = Object.assign(defaultCommands, commands);
```
Comment (deepu105) [inline cli/commands.js]: Use spread plz
```
module.exports = {
...defaultCommands,
...commands
}
```

**Comment by deepu105** [comment]
I'm ok with idea

File: cli/utils.js (hunk 1 above)
Comment (sahbi-ktifa) [reply cli/utils.js]: You are totally right !

File: cli/commands.js (hunk 2 above)
Comment (sahbi-ktifa) [reply cli/commands.js]: I'm not used to use this syntax :) 
Thanks for the hint!

**Comment by avdev4j** [comment]
Hi @sahbi-ktifa 
That's sound great since blueprints can't, always, match with the whole client need (call "the ultra specific part" :D).

I wonder if we should put the command file in a "cli" folder: "cli/commands.js" to match with the orignal JHipster file.

**Comment by sahbi-ktifa** [comment]
Hello @avdev4j, thanks for the review!
What did you meanusing the cli? What was your intent ?

**Comment by sahbi-ktifa** [comment]
Ok, I changed `commands.js` file under `cli` folder.
This is easier to understand and maintain.

File: cli/commands.js [hunk 3]
```diff
@@ -16,7 +16,15 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-module.exports = {
+
+let customCommands = {};
+if (process.argv.indexOf('--blueprint') > -1) {
```
[the remark this answers is not in the document]
Comment (sahbi-ktifa) [reply cli/commands.js]: Fixed it :)

**Pull request merged**